File: python/leetcode/stream_characters.py
class Trie:
    def __init__(self):
        self.children = [None] * 26
        self.isWord = False

    def insert(self, s):
        t = self
        for c in s:
            if not t.children[ord(c) - ord('a')]:
                t.children[ord(c) - ord('a')] = Trie()
            t = t.children[ord(c) - ord('a')]
        t.isWord = True

    def search(self, queries):
        t = self
        for c in queries:
            if  not t.children[ord(c) - ord('a')]: return False
            t = t.children[ord(c) - ord('a')]
            if t.isWord: return True
        return False

class StreamChecker(object):

    def __init__(self, words):
        """
        :type words: List[str]
        """
        self.trie = Trie()
        for s in words:
            self.trie.insert(s[::-1])
        self.queries = collections.deque()

    def query(self, letter):
        """
        :type letter: str
        :rtype: bool
        """
        self.queries.appendleft(letter)
        return self.trie.search(self.queries)

File: src/main/java/LeetCode/StreamCharacters.java
package LeetCode;

import java.util.*;

public class StreamCharacters {
    class Trie {
        public Trie[] children;
        public boolean isWord;

        public Trie() {
            children = new Trie[26];
            isWord = false;
        }

        public void insert(String s) {
            Trie t = this;
            for (char c : s.toCharArray()) {
                if (t.children[c-'a'] == null)
                    t.children[c-'a'] = new Trie();
                t = t.children[c-'a'];
            }
            t.isWord = true;
        }

        public boolean search(Deque<Character> queries) {
            Trie t = this;
            for (char c : queries) {
                if (t.children[c-'a'] == null) return false;
                t = t.children[c-'a'];
                if (t.isWord) return true;
            }
            return false;
        }
    }

    private static Trie trie;
    private static Deque queries;

    public StreamCharacters(String[] words) {
        trie = new Trie();
        for (String s : words) {
            String strRev = new StringBuilder(s).reverse().toString();
            trie.insert(strRev);
        }
        queries = new LinkedList();
    }

    public static boolean query(char letter) {
        queries.addFirst(letter);
        return trie.search(queries);
    }

    public static void main(String[] args) {
        String[] str = {"cd", "f", "kl"};
        StreamCharacters stream = new StreamCharacters(str);
        System.out.println(stream.query('a'));
        System.out.println(stream.query('b'));
        System.out.println(stream.query('c'));
        System.out.println(stream.query('d'));
        System.out.println(stream.query('e'));
        System.out.println(stream.query('f'));
    }
}
